implementations/rust/ockam/ockam_api/src/util.rs

```diff
@@ -127,29 +132,47 @@ pub fn multiaddr_to_addr(ma: &MultiAddr) -> Option<Address> {
 pub fn route_to_multiaddr(r: &Route) -> Option<MultiAddr> {
     let mut ma = MultiAddr::default();
     for a in r.iter() {
-        match a.transport_type() {
-            TCP => {
-                if let Ok(sa) = SocketAddrV4::from_str(a.address()) {
-                    ma.push_back(Ip4::new(*sa.ip())).ok()?;
-                    ma.push_back(Tcp::new(sa.port())).ok()?
-                } else if let Ok(sa) = SocketAddrV6::from_str(a.address()) {
-                    ma.push_back(Ip6::new(*sa.ip())).ok()?;
-                    ma.push_back(Tcp::new(sa.port())).ok()?
-                } else if let Some((host, port)) = a.address().split_once(':') {
-                    ma.push_back(DnsAddr::new(host)).ok()?;
-                    ma.push_back(Tcp::new(u16::from_str(port).ok()?)).ok()?
-                } else {
-                    ma.push_back(DnsAddr::new(a.address())).ok()?
-                }
-            }
-            LOCAL => ma.push_back(Service::new(a.address())).ok()?,
-            other => {
-                error!(target: "ockam_api", transport = %other, "unsupported transport type");
-                return None;
+        ma.try_extend(&try_address_to_multiaddr(a).ok()?).ok()?
+    }
+    Some(ma)
+}
+
+/// Try to convert an Ockam Address to a MultiAddr.
+pub fn try_address_to_multiaddr(a: &Address) -> Result<MultiAddr, Error> {
+    let mut ma = MultiAddr::default();
+    match a.transport_type() {
+        TCP => {
+            if let Ok(sa) = SocketAddrV4::from_str(a.address()) {
+                ma.push_back(Ip4::new(*sa.ip()))?;
+                ma.push_back(Tcp::new(sa.port()))?
+            } else if let Ok(sa) = SocketAddrV6::from_str(a.address()) {
+                ma.push_back(Ip6::new(*sa.ip()))?;
+                ma.push_back(Tcp::new(sa.port()))?
+            } else if let Some((host, port)) = a.address().split_once(':') {
+                ma.push_back(DnsAddr::new(host))?;
+                let n = u16::from_str(port).map_err(ApiError::wrap)?;
+                ma.push_back(Tcp::new(n))?
+            } else {
+                ma.push_back(DnsAddr::new(a.address()))?
             }
         }
+        LOCAL => ma.push_back(Service::new(a.address()))?,
+        other => {
+            error!(target: "ockam_api", transport = %other, "unsupported transport type");
+            return Err(ApiError::message(format!(
+                "unknown transport type: {other}"
+            )));
+        }
     }
-    Some(ma)
+    Ok(ma)
+}
+
+pub fn is_secure_channel_addr(a: &MultiAddr) -> bool {
```

Comment:
Does it check if secure channel will be created while using that MultiAddr, and doesn't detect if there is already existent secure channel in that MultiAddr? Would be nice to explain that in the comments (and maybe even in the name)

Reply:
Also wanted to clarify, is it possible that we want to check if next hop uses secure channel, e.g.
`/ipv4/127.0.0.1/tcp/1234/secure/api/ipv4/127.0.0.1/tcp4321/secure/api`, but for example this multiaddr also passes the check I guess? `/ipv4/127.0.0.1/tcp/1234/ipv4/127.0.0.1/tcp4321/secure/api`
I'm afraid of misusing that function

Reply:
To avoid confusion I have removed it in https://github.com/build-trust/ockam/pull/3528/commits/08d78aed923f83858d6996fcc98fa4a09f71a9a8. Reading the match inline seems to be more obvious to readers.